modules/contact/include/utils/MortarContactUtils.h

```diff
@@ -136,6 +160,202 @@ communicateVelocities(std::unordered_map<const DofObject *, T> & dof_map,
   TIMPI::push_parallel_vector_data(communicator, push_back_data, sent_back_action_functor);
 }
 
+/**
+ * This function is used to communicate velocities across processes
+ * @param dof_map_adr2t Map from degree of freedom to weighted tank two tensor
+ * @param mesh Mesh used to locate nodes or elements
+ * @param nodal Whether the element has Lagrange interpolation
+ * @param communicator Process communicator
+ * @param send_data_back Whether to send back data to a distributed constraint
+ */
+inline void
+communicateR2T(std::unordered_map<const DofObject *, ADRankTwoTensor> & dof_map_adr2t,
+               const MooseMesh & mesh,
+               const bool nodal,
+               const Parallel::Communicator & communicator,
+               const bool send_data_back)
+{
+  libmesh_parallel_only(communicator);
+  const auto our_proc_id = communicator.rank();
+
+  // We may have weighted velocity information that should go to other processes that own the dofs
+  using Datum = std::pair<dof_id_type, ADRankTwoTensor>;
+  std::unordered_map<processor_id_type, std::vector<Datum>> push_data;
+
+  for (auto & pr : dof_map_adr2t)
+  {
+    const auto * const dof_object = pr.first;
+    const auto proc_id = dof_object->processor_id();
+    if (proc_id == our_proc_id)
+      continue;
+
+    push_data[proc_id].push_back(std::make_pair(dof_object->id(), std::move(pr.second)));
+  }
+
+  const auto & lm_mesh = mesh.getMesh();
+  std::unordered_map<processor_id_type, std::vector<const DofObject *>>
+      pid_to_dof_object_for_sending_back;
+
+  auto action_functor =
+      [nodal,
+       our_proc_id,
+       &lm_mesh,
+       &dof_map_adr2t,
+       &pid_to_dof_object_for_sending_back,
+       send_data_back](const processor_id_type pid, const std::vector<Datum> & sent_data)
+  {
+    mooseAssert(pid != our_proc_id, "We do not send messages to ourself here");
+    libmesh_ignore(our_proc_id);
+
+    for (auto & pr : sent_data)
+    {
+      const auto dof_id = pr.first;
+      const auto * const dof_object =
+          nodal ? static_cast<const DofObject *>(lm_mesh.node_ptr(dof_id))
+                : static_cast<const DofObject *>(lm_mesh.elem_ptr(dof_id));
+      mooseAssert(dof_object, "This should be non-null");
+
+      if (send_data_back)
+        pid_to_dof_object_for_sending_back[pid].push_back(dof_object);
+
+      for (const auto i : make_range(3))
+        for (const auto j : make_range(3))
+          dof_map_adr2t[dof_object](i, j) += pr.second(i, j);
```

Comment:
This could be generalized nicely with the new  [`SerialAccess`](https://github.com/idaholab/moose/blob/next/framework/include/utils/SerialAccess.h) stuff ([unit test](https://github.com/idaholab/moose/blob/next/unit/src/SerialAccessTest.C)).

Reply:
I'll take a look.